### PR TITLE
only allow code to bind to the user's configured port numbers/ranges

### PR DESCRIPTION
**Pull request by rmuir**
Random code shouldn't be listening on sockets elsewhere.

Today its the wild west, but we only need to grant access to what the user configured.

This means e.g. multicast plugin has to declare its intentions in its security.policy

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/bootstrap/Security.java [hunk 1]
```diff
@@ -212,7 +226,36 @@ static Permissions createPermissions(Environment environment) throws IOException
...
+    }
+    
+    static void addBindPermissions(Permissions policy, Settings settings) throws IOException {
+        // http is simple
+        String httpRange = settings.get("http.netty.port", 
+                               settings.get("http.port", 
+                                       NettyHttpServerTransport.DEFAULT_PORT_RANGE));
+        policy.add(new SocketPermission("localhost:" + httpRange, "listen,resolve"));
+        // transport is waaaay overengineered
+        Map<String, Settings> profiles = settings.getGroups("transport.profiles", true);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: The code in NettyTransport does much more with it's settings. See for example NettyTransport.createFallbackSettings and how it's used . I agree what it does is probably an overkill, but can we make the methods there static and use the results setting here (using static strings for settings)?

**Comment by bleskes** [comment]
I think this is great (though I'm not an expert on the security config code). Left one comment ...

File: core/src/main/java/org/elasticsearch/bootstrap/Security.java (hunk 1 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: that code is not relevant. we don't care about addresses, only port numbers.

only the port number is checked. see http://docs.oracle.com/javase/8/docs/api/java/lang/SecurityManager.html#checkListen-int-

**Comment by rmuir** [comment]
I added a comment. just to be clear here, when binding, code just calls http://docs.oracle.com/javase/8/docs/api/java/lang/SecurityManager.html#checkListen-int- passing the port number. so the "localhost" you see in the permission is not really relevant (the actual address being bound to is not checked by java security, only the port number).

File: core/src/main/java/org/elasticsearch/bootstrap/Security.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I was referring to the fact that the setting netty uses are the result of a merge of multiple things:

```
 // merge fallback settings with default settings with profile settings so we have complete settings with default values
                    Settings mergedSettings = settingsBuilder()
                            .put(fallbackSettings)
                            .put(defaultSettings)
                            .put(profileSettings)
                            .build();
```

If we can make that static and available we can now for sure they use the same thing, even if people change stuff.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I don't want our security code entagled in this profile code.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: and for the record, if the parameters to configure which port we bind to are so complex, that we can't write readable logic to explain it, thats a much bigger problem. 

I insist on not entangling this with that broken code. I assigned https://github.com/elastic/elasticsearch/issues/14553 and will nuke it, despite any complaints to the contrary. its holding back everyone and breaking everything.

but it does not need to hold back improvements on security.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I kind of agree that we should separate the two. We don't do any default port stuff so all the stuff that transport does is really not relevant here. I wonder if we should make transport a hell lot more strict here and barf if ports are not explicit. That seems to me the right fix rather than pull the leniency into the security rule code.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I don't want to refactor transport in this PR, i already made separate issues for that.

The leniency is not pulled in here. Its not lenient, you don't specify port, we don't add permission -> access denied if you try to bind to that port.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: lets go with what we have and figure out what to do with transport in another PR / issue @bleskes are you ok with this?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I agree with the all the arguments around the transport profile handling code. What I'm saying is that we should reuse it (not copy it!) in Security to make sure we reach the same conclusions. This will also help prevent bugs when we fix the issues in the NettyTransport's settings code.

Here is what I meant : https://github.com/elastic/elasticsearch/commit/ac516f6b733b8ceddd9029fe87a0e12ceeb2aab1?w=1

IMHO, a good move while not violating @rmuir *valid* requirement not to pull the code into Security.java

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I simply don't agree its an improvement. There is no code reuse, that's because the netty transport code is doing a different thing when it iterates through the profiles.

Instead it just makes the security code reach out to more code (when its in an unprotected state) and stay less contained, which is a strong disadvantage.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: I think we should separate the two and push this as is. Your code refactoring has more changes than this functional change and on the security end I think we should be careful. let get this in and cleanup the stuff afterwards

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: In fact, even the constants used by my patch are dangerous: now they force NettyTransport to be loaded, look at its static block, which then loads NettyUtils and then netty classes... (god knows what they are doing, now and in the future)

Previously this stuff was all initialized after security manager was loaded, so we didn't have to worry about sheisty things happening.

I'm just mentioning this to explain my reasoning.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: > stay less contained, which is a strong disadvantage.
> I'm just mentioning this to explain my reasoning.

Thank you! 

I understand them and agree. My goal here is to make sure we do the same thing. For example, the current code doesn't check "http.enabled" and will enable port 9200 even if it's not needed (this specific example was very hard to catch because it wasn't even visible in the `NettyHttpServerTransport`). Same goes for `network.server` (which is probably bogus, but you get my point).

> Look at its static block, which then loads NettyUtils and then netty classes...

Argh. That's indeed very unfortunate.

I guess we should long term extract all of this to a NetworkSettings class of similar, but that is indeed a big refactoring and shouldn't be in the way of this.

I'm OK with proceeding with the current approach. Can I just ask that we make the settings keys static so we can use them to track all places where they are used? I'm afraid we'll make mistakes in future refactoring.

PS. I presume that just accessing a static const (like `NettyTransport.DEFAULT_PORT_RANGE` ) doesn't execute static coded?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: > PS. I presume that just accessing a static const (like `NettyTransport.DEFAULT_PORT_RANGE` ) doesn't execute static coded?

Correct. As long as the variable is marked `final`, is a primitive or a `String`, and is initialized by a compile-time constant expression then accessing it will not cause the class to be considered accessed and therefore not trigger the class variable initializers and static initializers to run ([JLS12.4.1](https://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.4.1), [JLS12.4.2](https://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.4.2)). So in this case,

```java
public static final String DEFAULT_PORT_RANGE = "9300-9400";
```

is okay and accessing it would not trigger `NettyUtils#setup` in the static initializer.

However, this strikes me as very fragile and subtle and I would want a minimum a test to ensure this behavior is preserved.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: Right, it gets baked in as a compile time constant in this particular case. Just mentioning I still see it as "playing with fire" and its still conceptually "reaching out" (someone innocently changes to a static method, now netty is doing something scary with Unsafe).

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: > However, this strikes me as very fragile and subtle and I would want a minimum a test to ensure this behavior is preserved.

That is why instead I would just prefer to keep the security code contained and not "reach out" to other code. Even if its at the expense of a few constants or helper methods. Thats just my opinion.

We know we make this tradeoff by allowing things like paths and ports to be configurable in elasticsearch.yml, the other option is to remove that flexibility, then we could initialize SM on JVM startup?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: hoped for a different answer as the previous code used `NettyTransport.DEFAULT_*`. Should we move these constants somewhere else then?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: > That is why instead I would just prefer to keep the security code contained and not "reach out" to other code. 

@rmuir Completely agree (I think I made a simultaneous edit to my [comment](https://github.com/elastic/elasticsearch/pull/14549#discussion_r44182190) to express this while you were writing the comment to which this is a reply).

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: > hoped for a different answer as the previous code used NettyTransport.DEFAULT_*. Should we move these constants somewhere else then?

These constants are OK, but in general it would be far better to have these constants somewhere else that does not also call methods in a static block. I see that as separate though. 

The current constant use will not cause any netty classes to be loaded. You can see this with -verbose:class

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: guys take this offline if you want - let move here.

**Comment by s1monw** [comment]
LGTM please push as is
